receiver/fluentforwardreceiver/conversion.go

```diff
@@ -0,0 +1,161 @@
+package fluentforwardreceiver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/internal/data"
+	"go.opentelemetry.io/collector/receiver/fluentforwardreceiver/protocol"
+	"go.uber.org/zap"
+)
+
+const tagAttributeKey = "fluent.tag"
+
+type Converter struct {
+	nextConsumer consumer.LogConsumer
+	eventCh      <-chan protocol.Event
+	logger       *zap.Logger
+}
+
+func newConverter(eventCh <-chan protocol.Event, next consumer.LogConsumer, logger *zap.Logger) *Converter {
+	return &Converter{
+		nextConsumer: next,
+		eventCh:      eventCh,
+		logger:       logger,
+	}
+}
+
+func (c *Converter) Start(ctx context.Context) {
+	go c.processEvents(ctx)
+}
+
+func (c *Converter) processEvents(ctx context.Context) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case e := <-c.eventCh:
+			buffered := []protocol.Event{e}
+			// Pull out anything waiting on the eventCh to get better
+			// efficiency on LogResource allocations.
+			buffered = fillBufferUntilChanEmpty(c.eventCh, buffered)
+
+			logs := convertEvents(buffered, c.logger)
+			c.nextConsumer.ConsumeLogs(ctx, logs)
+		}
+	}
+}
+
+func fillBufferUntilChanEmpty(eventCh <-chan protocol.Event, buf []protocol.Event) []protocol.Event {
+	for {
+		select {
+		case e2 := <-eventCh:
+			buf = append(buf, e2)
+		default:
+			return buf
+		}
+	}
+}
+
+func convertEvents(events []protocol.Event, logger *zap.Logger) data.Logs {
+	out := data.NewLogs()
+
+	logs := out.ResourceLogs()
+
+	logs.Resize(1)
+	rls := logs.At(0)
+
+	rls.Resource().InitEmpty()
+
+	logSlice := rls.Logs()
+
+	for _, event := range events {
+		switch ev := event.(type) {
+		case *protocol.MessageEvent:
+			lr := convertMessageEvent(ev)
+			logSlice.Append(lr)
+		case *protocol.ForwardEvent:
+			for i := range ev.Entries {
+				logSlice.Append(convertForwardEntry(ev.Tag, &ev.Entries[i]))
+			}
+		case *protocol.PackedForwardEvent:
+			entries, err := ev.Entries()
+			if err != nil {
+				logger.Error("Failed to get PackedForward entries", zap.Error(err))
+				continue
+			}
+
+			for i := range entries {
+				logSlice.Append(convertForwardEntry(ev.Tag, &entries[i]))
+			}
+		default:
+			panic("programmer forgot a case")
+		}
+	}
+
+	return out
+}
+
+func convertMessageEvent(me *protocol.MessageEvent) *pdata.LogRecord {
+	lr := pdata.NewLogRecord()
+	lr.InitEmpty()
+
+	if ts, err := me.Timestamp(); err == nil {
+		lr.SetTimestamp(pdata.TimestampUnixNano(ts.UnixNano()))
+	}
+
+	if logField, ok := me.Record["log"].(string); ok {
```

Comment:
Honestly I think if a `log` field comes through and it isn't a string it is kind of an unexpected scenario since we are specifically targeting `log` because of the strong convention of it being the log message from certain inputs to fluent-bit.